pkg/controllers/applicationfailover/rb_application_failover_controller.go

```diff
@@ -0,0 +1,247 @@
+
+	configv1alpha1 "github.com/karmada-io/karmada/pkg/apis/config/v1alpha1"
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	"github.com/karmada-io/karmada/pkg/features"
+	"github.com/karmada-io/karmada/pkg/resourceinterpreter"
+	"github.com/karmada-io/karmada/pkg/sharedcli/ratelimiterflag"
+	"github.com/karmada-io/karmada/pkg/util/fedinformer/keys"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+)
+
+// RBApplicationFailoverControllerName is the controller name that will be used when reporting events.
+const RBApplicationFailoverControllerName = "resource-binding-application-failover-controller"
+
+// RBApplicationFailoverController is to sync ResourceBinding's application failover behavior.
+type RBApplicationFailoverController struct {
+	client.Client
+	EventRecorder      record.EventRecorder
+	RateLimiterOptions ratelimiterflag.Options
+
+	// workloadUnhealthyMap records which clusters the specific resource is in an unhealthy state
+	workloadUnhealthyMap *workloadUnhealthyMap
+	ResourceInterpreter  resourceinterpreter.ResourceInterpreter
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *RBApplicationFailoverController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling ResourceBinding %s.", req.NamespacedName.String())
+
+	binding := &workv1alpha2.ResourceBinding{}
+	if err := c.Client.Get(ctx, req.NamespacedName, binding); err != nil {
+		if apierrors.IsNotFound(err) {
+			resource, err := helper.ConstructClusterWideKey(binding.Spec.Resource)
+			if err != nil {
+				return controllerruntime.Result{Requeue: true}, err
+			}
+			c.workloadUnhealthyMap.delete(resource)
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	if !c.bindingFilter(binding) {
+		resource, err := helper.ConstructClusterWideKey(binding.Spec.Resource)
+		if err != nil {
+			return controllerruntime.Result{Requeue: true}, err
+		}
+		c.workloadUnhealthyMap.delete(resource)
+		return controllerruntime.Result{}, nil
+	}
+
+	retryDuration, err := c.syncBinding(binding)
+	if err != nil {
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	if retryDuration > 0 {
+		klog.V(4).Infof("Retry to check health status of the workload after %v minutes.", retryDuration.Minutes())
+		return controllerruntime.Result{RequeueAfter: retryDuration}, nil
+	}
+	return controllerruntime.Result{}, nil
+}
+
+func (c *RBApplicationFailoverController) detectFailure(clusters []string, tolerationSeconds *int32, resource keys.ClusterWideKey) (int32, []string) {
+	var needEvictClusters []string
+	var duration int32
```

Comment:
What is this variable used for?

Reply:
This variable is always equal to `tolerationSeconds`, and `tolerationSeconds` has been constant. I haven't seen the purpose of this variable.

Reply:
This variable is not always equal to `tolerationSeconds`. When there is no need to detect after the tolerance time, duration is equal to 0.

Reply:
Do we need a logic like this if if `metav1.Now().After(unHealthyTimeStamp.Add(time.Duration(*tolerationSeconds) * time.Second))` equal to `false` ?


``` go
duration = *tolerationSeconds - int32(metav1.Now().Sub(unHealthyTimeStamp.Time).Seconds())
```

Reply:
When the fault is detected for the first time, it will rejoin the queue after tolerationSeconds. I haven't thought of a scenario where I need to rejoin the queue halfway.

Reply:
>  I haven't thought of a scenario where I need to rejoin the queue halfway.

 I think this is necessary.